Unify StorageVersion and SerializationVersion

ccfelius opened the pull request:
This PR unifies the `StorageVersion` and `SerializationVersion` since they have merged in recent DuckDB versions. It also changes hardcoded constants into semantically meaningful constants.

Starting with DuckDB `v1.5.0`, the legacy `serialization_version` (which utilized constants ranging from 1 to 7) will be phased out. This means, that where we formerly wrote the `serialization_version`, we align it with a `storage_version` and we now write a `storage_version` >= 68 (`v1.5.0)`.

We do keep the `serialization_version` around for backwards compatibility, which is specifically necessary for serialization version 1, 2 and 3, which all fall under storage version 64 (the default, < `v1.2.0`).

To  keep the storage version backwards compatible, we changed the `optional_idx storage_version` into a `StorageVersionMapping` struct:

```
struct StorageVersionMapping {
    optional_idx version;
    string version_string;
};
```

When a database version is read, we now determine the _semantically_ corresponding storage version by (i) for versions `<v1.5.0` converting the legacy `serialization_version` into its associated version string and (ii) for versions `>v1.5.0` we convert the serialized `storage_version` into its associated string.

In code paths where explicit version checks are required for any serialization version prior to version 4  / 64 (`<v1.2.0`), we look at the version string, which is input for the `GetSerializationVersionDeprecated()` method. The returned deprecated serialization constant is then used to facilitate comparisons against these old serialization versions.

Fixes https://github.com/duckdblabs/duckdb-internal/issues/5766.

Feedback very welcome @carlopi & @Tishj

Commit:
changing SerializationCompatbility to unify storage versions

# Conflicts:
#	src/storage/storage_info.cpp
#	src/storage/version_map.json

Commit:
change script and replace const

Commit:
add comment

Commit:
convert storage numbers

Commit:
wip

Commit:
wip

Commit:
replacing empty dbs

Commit:
adding deprecated serialization enum

Commit:
add serializationversiondepricated to exclude enum

Commit:
adding test

Commit:
remove outcommented code

Commit:
change constants for semantical alternatives

Commit:
fix tidy error

Commit:
generate storage serialization version:

# Conflicts:
#	test/sql/storage_version/storage_version.db

Commit:
fix deprecated serialization versions

Commit:
debug storage version 1.4.2

fix failing tests + generated code

remove comments

remove comment

Commit:
recovering tests

Commit:
adding v1.4.0 db file

Commit:
adding missing database

carlopi commented:
I have some doubts on how to roll this in connected to maintaining multiple branches and whether it's viable to uniform given:
```
	{"v1.0.0", 2},
	{"v1.1.0", 3},
```
vs
```
	{"v1.0.0", 64},
	{"v1.1.0", 64},
```
So basically, v1_0_0 and v1_1_0 afterward will both go to 64, but it's not an issue that those can't be told apart?
In particular, connected to code at https://github.com/duckdb/duckdb/pull/20288/changes#diff-3db3d0ea8aa71c059403034552d5ab3b20e1300cda092dcd8407a480cbdf07d4R121-R122

Commit:
using serialization version for v2 and v3

Commit:
Add StorageVersionMapping

Commit:
wip

Commit:
fixing bugs and adding semantic ser version

Commit:
remove unnecessary serialization

Commit:
update generate serialization script:

Commit:
small fixes

Commit:
fix for forward compatibility

Commit:
Format fix

Commit:
make CI pass

string_t -> std string

tidy fix; pass by reference

fixing assert

remove faulty test

add missing database

fix string inline test

enter

Commit:
add v string for latest

Commit:
change hidden loop for constant

change getvalue func to static cast

last optimization

change missing strings to constants

ccfelius commented:
@carlopi completely refactored the PR, now it should cover all the cases. Feel free to take another pass on this :)

carlopi commented:
I just tried something like:
```
carlo@Mac duckdb % duckdb my_randomly_named.db
DuckDB v1.4.3 (Andium) d1dc88f950
Enter ".help" for usage hints.
D 
D 

carlo@Mac duckdb % ./build/release/duckdb     
DuckDB v1.5.0-dev5314 (Development Version, 502732cae1)
Enter ".help" for usage hints.
memory D ATTACH 'my_randomly_named.db' AS x (STORAGE_VERSION 'v1.5.0');
memory D CREATE TABLE x.T AS SELECT 32;
memory D 
memory D 

carlo@Mac duckdb % ./build/release/duckdb
DuckDB v1.5.0-dev5314 (Development Version, 502732cae1)
Enter ".help" for usage hints.
memory D ATTACH 'my_randomly_named.db' AS x (STORAGE_VERSION 'v1.5.0');
Invalid Input Error:
Deprecated Serialization Version is not found!
memory D 
```
where `duckdb` is the `brew` installed v1.4.3 binary and `/build/release/duckdb` built on the bases of your branch.

Also, unsure if this helps, but changing back to `v1.5-variegata`, and rebuilding, I get:
```
memory D ATTACH 'my_randomly_named.db' AS x (STORAGE_VERSION 'v1.5.0');
Invalid Input Error:
Error opening "my_randomly_named.db": cannot initialize database with storage version 7 - which is lower than what the database itself uses (68). The storage version of an existing database cannot be lowered.
memory D 
memory D 

carlo@Mac duckdb % ./build/release/duckdb
DuckDB v1.5.0-dev5324 (Development Version, 483fbf12c5)
Enter ".help" for usage hints.
memory D ATTACH 'my_randomly_named.db' AS x;
Invalid Input Error:
Error opening "my_randomly_named.db": file was written with a storage version greater than the latest version supported by this DuckDB instance. Try opening the file with a newer version of DuckDB.
memory D 
memory D 
```

Could you check if you reproduce?
I have not yet looked at the code, just testing different things.

Commit:
fix for bumping to v1.5.0

Commit:
add newly generated database file

carlopi commented on src/storage/single_file_block_manager.cpp:
```diff
@@ -666,27 +739,31 @@ void SingleFileBlockManager::Initialize(const DatabaseHeader &header, const opti
+			    "Error opening \"%s\": cannot initialize database with storage version %d (%s) - which is lower than "
+			    "what the database itself uses (%d). The storage version of an existing database cannot be lowered.",
+			    path, requested_compat_version, header.storage_compatibility.version_string,
+			    header.storage_compatibility.version.GetIndex());
```
This message is off, I think it should be:
```
Error opening \"%s\": cannot initialize database with storage version %d - which is lower than "
			    "what the database itself uses (%d - %s). The storage version of an existing database cannot be lowered.
```
while swapping params.
If you want to provide a string description of the first number, you need likely to find that in the map.

carlopi reviewed (commented):
I think this is basically ready, error message could use fixing up, since now name refers to the wrong version, but that's local.

This is a quite scary PR to merge, but looks about right, thanks!

Could you retarget to `v1.5-variegata` branch? Thanks!

Commit:
change error message, add test

Commit:
fix windows failure

Mytherin commented on src/include/duckdb/storage/compression/alp/alp_analyze.hpp:
```diff
@@ -36,7 +36,7 @@ struct AlpAnalyzeState : public AnalyzeState {
 	vector<vector<T>> complete_vectors_sampled;
 	alp::AlpCompressionData<T, true> compression_data;
-	idx_t storage_version = 0;
+	idx_t storage_version = static_cast<idx_t>(StorageVersion::INVALID);
```
Maybe this should be an `optional_idx` in that case?

Mytherin reviewed (commented):
Thanks for the PR! Looks good - some comments. Unless I'm missing something I would vote for removing the serialization version entirely, however. I think that will simplify this PR significantly as well.

carlopi commented on src/include/duckdb/storage/storage_info.hpp:
```diff
@@ -64,15 +66,161 @@ struct Storage {
+// END OF ENUM VERSION INFO
+
+// START OF SER_ENUM VERSION INFO
+enum class SerializationVersionDeprecated : uint64_t {
```
| Maybe @carlopi has an opinion here as well?

On `SerializationVersionDeprecated`, I agree it could be cleaned up, I would consider, but for @ccfelius to decide, that maybe it could be a follow up PR? Meaning goal here is restructuring while keeping parity, next PR is cleaning up AND forbid those as targets?

---

On the MainHeader vs DatabaseHeader, I would keep that as of now, but given you have it now in your head some extra comments / explainer would be also super handy for future versions of us.
